Fix bug when decrementing the user connection count

AndrewJackson2020 opened the pull request:
Fixes #1237

The user connection count was not being decremented correctly if the user did not exist in the auth file, but did exist in the pgbouncer config.

Relevant docs:

> Note that when auth_file is configured, if a user is defined in this section but not listed in auth_file, pgBouncer will attempt to use auth_query to find a password for that user if auth_user is set. If auth_user is not set, pgBouncer will pretend the user exists and fail to return "no such user" messages to the client, but neither will it accept any provided password.

Commit:
Add test to recreate error

JelteF commented:
Would a fix similar to this work: https://github.com/pgbouncer/pgbouncer/pull/1166

Commit:
Fix

AndrewJackson2020 commented:
> Would a fix similar to this work: #1166

Was able to fix this by adding a call to find_or_add_new_global_user.

AndrewJackson2020 commented:
Any chance we can merge this in before the next release (assuming the fix looks reasonable)?

AndrewJackson2020 commented:
> I didn't check this PR but if it is a recent issue, I would say, fix it before release. If it is an old issue, we might postpone it for the next release. Of course, we also need to consider how critical it is. Since it is just a wrong number, it is not a critical one. I would say, fix it if we are sure the patch is correct. I can take a look in a couple of hours.

This is a fix related to #1137 which has not made it to a release yet (new functionality). Also if someone tries to use max_user_client_connections this will cause users to not be able to connect, so in that case its a bit more critical than an incorrect number.

JelteF commented:
Yeah agreed, planning to look at this soonish. Test looks good, I'm mainly wondering if the fix is in the correct spot.

JelteF committed:
Make the test clearer and add an extra test

The additional test already passed before but seems like a good test
case to handle too.

AndrewJackson2020 commented:
Okay looking at your test my fix just shuffled the issues around a bit. Will look into this.

JelteF commented:
Hmm, the additional test passed locally on my machine I believe. I might have messed something up right before pushing it.

Commit:
Fix

AndrewJackson2020 commented:
I believe that I fixed it with an if statement and a call to `find_global_user` instead. This way non-existent users aren't being created but users that were created because they had entries in the conf file are assigned. Possibly there is a better way of doing this though.

JelteF committed:
Add some comments

JelteF reviewed (approved):
Added a few comments for clarity. I think this is good now.

JelteF committed:
Extra word

AndrewJackson2020 committed:
Merge branch 'master' into user_decrement_error

AndrewJackson2020 commented:
Merged with main branch. All tests green. lgtm
